Make the error more strongly typed

gkoz opened the pull request:
Allow users to distinguish between pkg-config being disabled by an environment variable, missing and genuinely failing.

gkoz commented:
Oops, it had slipped my mind that enum variants can have named fields, I'll fix that tomorrow.

alexcrichton commented:
Hm, this is unfortunately a breaking change, which would require a major version bump of this library. I think I'd prefer to do this through "pseudo deprecation" like:

* Add a new function (maybe named something like "probe") which sits next to the existing ones
* Implement the old functions in terms of the new
* Mark the old functions as `#[doc(hidden)]` and document them as deprecated

In terms of backwards compatibility of the `enum` itself I may want to think a bit more about the variants and the payloads, but I think that using named fields would probably help more here, yeah. Another thing to do would perhaps be to add a `#[doc(hidden)]` `__Nonexhaustive` variant documented as saying "please don't match on this, we're likely to add more variants over time".

gkoz commented:
Why is major version bump undesirable?

I've addressed the comments.

The [lack of `impl Debug for Output` in stable libstd](https://travis-ci.org/alexcrichton/pkg-config-rs/jobs/109661704#L151) might suggest my choice of error field types is suboptimal. Alternatively we could wait for 1.7 and do bump the version.

alexcrichton commented:
A major version bump is basically undesirable because it's a major version bump. Most users probably won't even see breakage, and with 99 reverse dependencies there's not really any reason to require an update.

Also yes, let's use an explicit `impl Debug` to work around the lack of impl for `Output` for now.

gkoz commented:
Added explicit `impl Debug` and version bump.

alexcrichton commented:
Thanks!

gkoz commented:
I'd also like to make `println!("cargo:...")` optional. Would you accept such feature and does it make sense to fit into the same 0.3.7 crate push?

alexcrichton commented:
Yes I can add that
